Add `grad_2F1` gradient for a2

spinkney opened the pull request:
## Summary

Adds gradient calculation for a2 input and fixes incorrect return values when `z` is between -1 and 0. Closes #2663.

## Release notes

Adds gradient calculation for the a2 input in stan-math for the `grad_2F1` function and fixes incorrect return values when `z` is between -1 and 0. 

## Checklist

- [x] Math issue #2663

- [x] Copyright holder: Sean Pinkney

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

spinkney commented:
@SteveBronder @sakrejda or @andrjohns this is ready for review

SteveBronder commented on stan/math/prim/fun/grad_2F1.hpp:
```diff
@@ -8,10 +8,60 @@
+ * @param[in] log_t_old log of previous power series calculation
+ * @param[in] log_t_old_sign log of previous power series calculation
+ * @param[in] k loop iteration
+ * @param[in] a1 a1 see hypergeometric function 2F1 definition
+ * @param[in] a2 a2 see hypergeometric function 2F1 definition
+ * @param[in] b1 b1 see hypergeometric function 2F1 definition
```
duplicate names

SteveBronder commented on stan/math/prim/fun/grad_2F1.hpp:
```diff
@@ -8,10 +8,60 @@
+ * @param[in] a2 a2 see hypergeometric function 2F1 definition
+ * @param[in] b1 b1 see hypergeometric function 2F1 definition
+ */
+template <typename T, typename T_array, typename T_array_int, typename T_int>
```
`std::array` and stan arrays are two very different things. I like the having more verbose names but I'd call these like `StdVec` and `StdVecInt`. Also note we use CamelCase for template parameter names.

SteveBronder commented on stan/math/prim/fun/grad_2F1.hpp:
```diff
@@ -24,82 +74,126 @@ namespace math {
+ * @tparam T2 scalar or `var_value`
+ * @tparam T3 scalar or `var_value`
+ * @tparam T_z scalar type
+ * @param[out] g_a1 g_a1 reference to gradient of 2F1 w.r.t. a1, result
+ * @param[out] g_a2 g_a2 reference to gradient of 2F1 w.r.t. a2, result
+ * @param[out] g_b1 g_b1 reference to gradient of 2F1 w.r.t. b1, result
+ * @param[in] a1 a1 see generalized hypergeometric function definition
+ * @param[in] a2 a2 see generalized hypergeometric function definition
+ * @param[in] b1 b1 see generalized hypergeometric function definition
+ * @param[in] z z see generalized hypergeometric function definition
```
duplicate names in docs

SteveBronder commented on stan/math/prim/fun/grad_2F1.hpp:
```diff
@@ -24,82 +74,126 @@ namespace math {
   using std::fabs;
   using std::log;
   using std::max;
+  using TP = return_type_t<T1, T2, T3, T_z>;
```
We use `_t` to and snake case to signify a type returned by a type trait

```suggestion
  using ret_t = return_type_t<T1, T2, T3, T_z>;
```

SteveBronder commented on stan/math/prim/fun/grad_2F1.hpp:
```diff
@@ -24,82 +74,126 @@ namespace math {
+              int max_steps = 1e6) {
   check_2F1_converges("grad_2F1", a1, a2, b1, z);
 
   using stan::math::value_of_rec;
   using std::exp;
   using std::fabs;
   using std::log;
   using std::max;
```
I know you didn't do this but personally I find it leads to headaches when we bring things into a local scope from `std`. imo I'd rather just make the longer call. Unless this function only works with doubles I think we should be using the stan math versions

SteveBronder commented on test/unit/math/prim/fun/grad_2F1_test.cpp:
```diff
@@ -111,37 +155,43 @@ TEST(MathPrimScalFun, grad2F1_9) {
+  EXPECT_NEAR(36347869.41885337, grad_a2, 1e-1);
   EXPECT_NEAR(-30843032.10697079073015067426929807, grad_b1,
               1e-1);  // reference: discrete diff in mathematica
 }
```
Should we be using the autodiff testing framework for this?

spinkney replied:
I don't think we can because we don't have the 2F1 function, this is - a bit confusingly because it's in prim - the gradient of 2F1.

SteveBronder reviewed (commented):
I'm just leaving a comment with some basic things. I don't know much about these functions so I think someone like @andrjohns would need to do that main review. Though if he is too busy I can try to sort out more of what's going on here

spinkney commented:
@andrjohns or @sakrejda this is ready for review

spinkney commented:
feature freeze is coming in 10 days. I'd like to get this in before then. @andrjohns can you review by then or do you have any suggestions for someone who could?

spinkney commented:
@andrjohns I simplified the code and I think this should be good to go. Mind taking a look?

andrjohns commented on stan/math/prim/fun/grad_2F1.hpp:
```diff
@@ -8,10 +8,60 @@
+  log_t_new += log(fabs(p)) + log_z;
+  log_t_new_sign = p >= 0.0 ? log_t_new_sign : -log_t_new_sign;
+
+  ret_t term
+      = log_g_old_sign[0] * log_t_old_sign * exp(log_g_old[0] - log_t_old)
+        + inv(a1 + k);
+  log_g_old[0] = log_t_new + log(fabs(term));
+  log_g_old_sign[0] = term >= 0.0 ? log_t_new_sign : -log_t_new_sign;
+
+  term = log_g_old_sign[1] * log_t_old_sign * exp(log_g_old[1] - log_t_old)
+         + inv(a2 + k);
+  log_g_old[1] = log_t_new + log(fabs(term));
+  log_g_old_sign[1] = term >= 0.0 ? log_t_new_sign : -log_t_new_sign;
+
+  term = log_g_old_sign[2] * log_t_old_sign * exp(log_g_old[2] - log_t_old)
+         - inv(b1 + k);
+  log_g_old[2] = log_t_new + log(fabs(term));
+  log_g_old_sign[2] = term >= 0.0 ? log_t_new_sign : -log_t_new_sign;
```
Given that the same calculations are being applied across the different arguments, we could probably vectorise this using something like:
```cpp
  const auto& log_g_old_vec = as_array_or_scalar(log_g_old);
  const auto& log_g_old_sign_vec = as_array_or_scalar(log_g_old_sign);
  Eigen::Array<T, Eigen::Dynamic, 1> hyper_args(3);
  hyper_args << a1, a2, b1;

  Eigen::Array<T, Eigen::Dynamic, 1> term
    = log_g_old_sign_vec * log_t_old_sign * exp(log_g_old_vec - log_t_old)
       + inv(hyper_args + k);

  log_g_old_vec = log_t_new + log(fabs(term));
  log_g_old_sign_vec = sign(value_of_rec(term)) * log_t_new_sign;

```

andrjohns reviewed (changes requested):
Thanks for putting this all together, it's all looking great! Just a couple of suggestion for efficiency, but otherwise just about ready to go!

spinkney commented:
@andrjohns thanks for the review! I've addressed the comments and I was able to update the lambda function for vectorization. I wasn't able to get `as_array_scalar` to work so instead I created the Eigen arrays outside of the lambda.

andrjohns reviewed (approved):
LGTM!
